### DISCUSSION TITLE
Is it possible to use CO-RE with kernel 4.9?

### DISCUSSION THREAD

**Discussion by tanshihaj** [cilium/ebpf#1033] [Q&A] [answered]
README.md states that cilium/ebpf support linux kernels 4.9+. However I cannot understand how applyRelocations() should work - BTF supported only form linux kernel 4.18.

Is there any way to load BPF program on 4.9 kernel?

**Comment by lmb** [comment, marked as answer]
So, `bpf2go` is a way for you to compile C to BPF and ship it in your go binary. It doesn't have a kernel dependency.

When you say `applyRelocations`, I'm assuming you're talking about CO-RE? The library allows you to provide your own BTF using [ProgramOptions.KernelTypes](https://pkg.go.dev/github.com/cilium/ebpf#ProgramOptions). You can combine this with [btfhub](https://github.com/aquasecurity/btfhub) which has BTF for popular distros / older kernels.

**Reply by tanshihaj** [reply]
> When you say applyRelocations, I'm assuming you're talking about CO-RE?

Yes, it seems my issue is connected with CO-RE.

> You can combine this with [btfhub](https://github.com/aquasecurity/btfhub) which has BTF for popular distros / older kernels.

Yes, it seems that I need to generate BTF for my 4.9 kernel somehow and pass it with ProgramOptions.

**Reply by tanshihaj** [reply]
It seems that latest (31bc0d7410572f6e03e3ed9da7c8c6f0d8df23c8) pahole from https://github.com/acmel/dwarves generates BTF from vmlinux for kernel 4.9 with command `./dwarves/build/pahole --btf_encode_detached vmlinux-4.9.253-....btf /boot/vmlinux-4.9.253-...`.
